sycl/source/detail/pi.cpp

```diff
@@ -197,7 +197,12 @@ bool trace(TraceLevel Level) {
 
 // Initializes all available Plugins.
 vector_class<plugin> initialize() {
-  vector_class<plugin> Plugins;
+  static bool PluginsInitDone = false;
+  static vector_class<plugin> Plugins;
```

Comment:
Who owns plugin? I thought it was platform_impl::MPlatform.
Related question:
What is the lifetime of Plugins? What happens if user creates a platform(to enter this function) from a global object destructor?

Reply:
plugins are system-wide, think of them as a global list of shared libraries loaded. it is currently created at the first call to pi::initialize() which I think (TODO) we should fine a better place for than in platforms_get(), and also (TODO) deallocate them at global tear-down.

Reply:
1. Still, who owns plugins right now? I see that platform_impl has `std::shared_ptr<plugin> MPlugin;` field, so I expect that platform_impl objects own plugins. When all platform_impl objects are destroyed plugin object will be deallocated as well, so `Plugins` vector will point to already released memory.

2. What happens if user creates a platform(to enter this function) from a global object destructor? For example the following SYCL application can(it's UB) catch a crash with your implementation:

```
class Glob {
~Glob() { sycl::platform Platform(...); }
}
Glob GVar;

int main() {
  sycl::platform Platform(...);
}

```

Reply:
I also agree there is a problem caused by undefined order of global objects destruction. The pointer solution is an interesting one (especially for its simplicity), an is probably a suitable one where destructors have no important side effects.

An alternative would be to put some order to the objects destruction (there are C++ extensions to do so, e.g. GNU attribute "destructor" with priority: https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html). Vlad mentioned there is already some llvm class like GlobalGlobal (not remembering the exact name), that are guaranteed to be destroyed after other globals.

Reply:
There are two disadvantages to using the attributes which control destructor ordering:

* According to the GCC documentation, those attributes are only supported on some targets (I'm not sure if this is also the case for the LLVM implementation).

* The application, of course, could also use this attribute to cause its global destructor to run even after ours, thus causing an "arms race" of destructor ordering.

I agree that the pointer solution does not work if there is some important action in the destructor that must be done before the application terminates.  I suspect this is probably not likely, though, for our implementation in the SYCL runtime.

Reply:
A downside of never deleted plugins is that valgrind and other sanitizers will keep telling us that we have a leak (we won't be able to get clean output from these very good tools). @romanovvlad , what do you think?

Reply:
I would prefer that we do not use global objects.

If it's really needed I would suggest having a global pointer solution which is deallocated in "GNU attribute "destructor"". 

> an is probably a suitable one where destructors have no important side effects.

Are you sure "plugins" have no important side effects? Shouldn't they at least dlclose a plugin library?

Reply:
> I would suggest having a global pointer solution which is deallocated in "GNU attribute "destructor""

Can we work on it separately from this change? 
I added a TODO explaining that we need to destruct plugins last of all.

Reply:
Ok.

Reply:
> Are you sure "plugins" have no important side effects? Shouldn't they at least dlclose a plugin library?

I don't think the `dlclose` is necessary.  Remember, this is happening right before the process terminates.  The library will get implicitly dlclosed when the process termintes.

Reply:
User can load libsycl.so using dlopen, pugins libraries will not be properly freed by dlclose.